pennylane/device_tracker.py

```diff
@@ -0,0 +1,171 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+from collections import defaultdict
+
+import pennylane as qml
+
+
+def track(dev, version="default", **kwargs):
+    r"""Creates a tracking context and applies it to a device.
+
+    Args:
+        dev (~.Device): a PennyLane-compatible device
+        version (str): name of tracker to use.  The current options are
+            `default` and `timing`.
+
+    Keyword Args:
+        reset_on_enter=True (bool): whether or not to reset information
+            entering the context
+
+    **Usage Information**
+
+    Note that with backpropagation, this functions should take ``qnode.device``
+    instead of the device used to create the QNode.
+
+    .. code-block:: python
+
+        dev = qml.device('default.qubit', wires=1)
+
+        @qml.qnode(dev, diff_method="parameter-shift")
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+    With the default version, total execution information is printed on
+    each device execution.  The printed data depends on the device and tracker version,
+    but for standard PennyLane devices, the object will track executions and shots.
+
+    >>> with qml.track(circuit.device) as tracker:
+    ...    qml.grad(circuit)(0.1, shots=10)
+    Total: executions = 1	shots = 10
+    Total: executions = 2	shots = 20
+    Total: executions = 3	shots = 30
+
```

Comment:
Suggest to put the `.. UsageDetails::` here, and make the above an `**Example**`